Mathlib/Data/Fintype/Basic.lean

```diff
@@ -155,14 +155,93 @@ theorem codisjoint_right : Codisjoint s t ↔ ∀ ⦃a⦄, a ∉ t → a ∈ s :
   Codisjoint_comm.trans codisjoint_left
 #align finset.codisjoint_right Finset.codisjoint_right
 
-section BooleanAlgebra
-
-variable [DecidableEq α] {a : α}
-
-instance booleanAlgebra : BooleanAlgebra (Finset α) :=
+instance booleanAlgebra [DecidableEq α] : BooleanAlgebra (Finset α) :=
   GeneralizedBooleanAlgebra.toBooleanAlgebra
 #align finset.boolean_algebra Finset.booleanAlgebra
 
+end Finset
+
+namespace Mathlib.Meta
+open Lean Elab Term Meta Batteries.ExtendedBinder
+
+/-- Elaborate set builder notation for `Finset`.
+
+* `{x | p x}` is elaborated as `Finset.filter (fun x ↦ p x) Finset.univ` if the expected type is
+  `Finset ?α`.
+* `{x : α | p x}` is elaborated as `Finset.filter (fun x : α ↦ p x) Finset.univ` if the expected
+  type is `Finset ?β`.
+* `{x ∉ s | p x}` is elaborated as `Finset.filter (fun x ↦ p x) sᶜ` if either the expected type is
+  `Finset ?α` or the expected type is not `Set ?α` and `s` has expected type `Finset ?α`.
+* `{x ≠ a | p x}` is elaborated as `Finset.filter (fun x ↦ p x) {a}ᶜ` if the expected type is
+  `Finset ?α`.
+
+See also
+* `Init.Set` for the `Set` builder notation elaborator that this elaborator partly overrides.
+* `Data.Finset.Basic` for the `Finset` builder notation elaborator partly overriding this one for
+  syntax of the form `{x ∈ s | p x}`.
+* `Data.Fintype.Basic` for the `Finset` builder notation elaborator handling syntax of the form
+  `{x | p x}`, `{x : α | p x}`, `{x ∉ s | p x}`, `{x ≠ a | p x}`.
+* `Order.LocallyFinite.Basic` for the `Finset` builder notation elaborator handling syntax of the
+  form `{x ≤ a | p x}`, `{x ≥ a | p x}`, `{x < a | p x}`, `{x > a | p x}`.
+
+TODO: Write a delaborator
+-/
+@[term_elab setBuilder]
+def elabFinsetBuilderSetOf : TermElab
+  | `({ $x:ident | $p }), expectedType? => do
+    -- AS we want to reason about the expected type, we would like to wait for it to be available.
+    -- However this means that if we fall back on `elabSetBuilder` we will have postponed.
+    -- This is undesirable as we want set builder notation to quickly elaborate to a `Set` when no
+    -- expected type is available.
+    -- tryPostponeIfNoneOrMVar expectedType?
+    -- If the expected type is not known to be `Finset ?α`, give up.
+    unless ← knownToBeFinsetNotSet expectedType? do throwUnsupportedSyntax
+    elabTerm (← `(Finset.filter (fun $x:ident ↦ $p) Finset.univ)) expectedType?
```

Comment:
This code is very repetitive, especially the comments. Do you think you can factor out a function (that holds all these comments), and then makes the `elabTerm` a straightforward application?

Reply:
How much should I factor out? All elaborators check `expectedType?`, but the elaborators for `{x \in s | p x}` and `{x \nin s | p x}` also check the type of `s`.

Further, I would think I should leave the comments in the end code because this is the one people will try modifying in the future.

In short, I don't think I can factor much more than the two lines `unless ... throwUnsupportedSyntax`.

Reply:
Actually I could just move the comment to `knownToBeFinsetNotSet` and it all makes sense